// express-simple-proxy/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = exports.DEFAULT_RETRY_COUNT = exports.MAX_REQUEST_SIZE = exports.DEFAULT_TIMEOUT = exports.asyncWrapper = exports.generateCurlCommand = exports.createFormDataPayload = exports.buildQueryString = exports.replaceUrlTemplate = exports.urlJoin = exports.defaultErrorHandler = exports.axiosProxyRequest = exports.createProxyController = void 0;
// Main exports
var proxy_1 = require("./proxy");
Object.defineProperty(exports, "createProxyController", { enumerable: true, get: function () { return proxy_1.createProxyController; } });
Object.defineProperty(exports, "axiosProxyRequest", { enumerable: true, get: function () { return proxy_1.axiosProxyRequest; } });
Object.defineProperty(exports, "defaultErrorHandler", { enumerable: true, get: function () { return proxy_1.defaultErrorHandler; } });
// Utility functions
var utils_1 = require("./utils");
Object.defineProperty(exports, "urlJoin", { enumerable: true, get: function () { return utils_1.urlJoin; } });
Object.defineProperty(exports, "replaceUrlTemplate", { enumerable: true, get: function () { return utils_1.replaceUrlTemplate; } });
Object.defineProperty(exports, "buildQueryString", { enumerable: true, get: function () { return utils_1.buildQueryString; } });
Object.defineProperty(exports, "createFormDataPayload", { enumerable: true, get: function () { return utils_1.createFormDataPayload; } });
Object.defineProperty(exports, "generateCurlCommand", { enumerable: true, get: function () { return utils_1.generateCurlCommand; } });
Object.defineProperty(exports, "asyncWrapper", { enumerable: true, get: function () { return utils_1.asyncWrapper; } });
// Constants
var types_1 = require("./types");
Object.defineProperty(exports, "DEFAULT_TIMEOUT", { enumerable: true, get: function () { return types_1.DEFAULT_TIMEOUT; } });
Object.defineProperty(exports, "MAX_REQUEST_SIZE", { enumerable: true, get: function () { return types_1.MAX_REQUEST_SIZE; } });
Object.defineProperty(exports, "DEFAULT_RETRY_COUNT", { enumerable: true, get: function () { return types_1.DEFAULT_RETRY_COUNT; } });
// Default export
var proxy_2 = require("./proxy");
Object.defineProperty(exports, "default", { enumerable: true, get: function () { return proxy_2.createProxyController; } });
